### PR TITLE
[MSINTEROP-722] Populate qe-metrics database with issues from provided Jira queries

### PR DESCRIPTION
**Pull request by calebevans**
- Adds logic to `cli.py` to execute each Jira query and populate the database accordingly.
- Removes the "unique" requirements from the `issue_key` field in the `JiraIssues` table in the database.
    - It was determined that this should be done because a bug can affect more than one "product" or "service" as discussed in the weekly qe-metrics meeting on 03/27.
- Adds a `date_record_modified` column to the `JiraIssues` table in the database for use in the data retention policy
- Adds class method `create_update_issues` to `JiraIssues`. Used to create or update records in the `JiraIssues` table from a list of issues returned from a query.
- Adds use of a context manager for the `Jira` class.
- Adds `issu_utils.py` to hold various supporting functions related to issue creation, updates, and deletion.
- Adds unit tests

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>


</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: qe_metrics/libs/database.py [hunk 1]
```diff
@@ -127,3 +134,44 @@ class JiraIssues(DB_CONNECTION.Entity):  # type: ignore
...
+            Args:
+                issues (List[Issue]): A list of Jira issues
+                product (Database.Products): A product object
+                severity (str): Severity of the issues
+                jira_server (str): Jira server URL
+
+            Returns:
+                List["Database.JiraIssues"]: A list of JiraIssues objects
+            """
+            issues = [_issue for _issue in issues if _issue.fields.issuetype.name.lower() == "bug"]
```
Comment (myakove) [inline qe_metrics/libs/database.py:155]: Don't we process all issues that we got from the query? 
The user needs to have a proper query.

File: qe_metrics/libs/database.py [hunk 2]
```diff
@@ -127,3 +134,44 @@ class JiraIssues(DB_CONNECTION.Entity):  # type: ignore
...
+            for issue in issues:
+                existing_issue = cls.get(issue_key=issue.key, product=product)
+                if existing_issue:
+                    update_existing_issue(existing_issue, issue, severity)
```
Comment (myakove) [inline qe_metrics/libs/database.py:159]:
```suggestion
                if existing_issue := cls.get(issue_key=issue.key, product=product):
                    update_existing_issue(existing_issue, issue, severity)
```

File: qe_metrics/libs/database.py [hunk 3]
```diff
@@ -127,3 +134,44 @@ class JiraIssues(DB_CONNECTION.Entity):  # type: ignore
...
+                        url=f"{jira_server}/browse/{issue.key}",
+                        project=issue.fields.project.key,
+                        severity=severity,
+                        status=issue.fields.status.name,
+                        customer_escaped=check_customer_escaped(issue=issue),
+                        date_created=format_issue_date(issue.fields.created),
+                        last_updated=format_issue_date(issue.fields.updated),
+                    )
+
+            delete_closed_issues(
```
Comment (myakove) [inline qe_metrics/libs/database.py:174]: Do we want to delete closed issues or leave it there and it will be deleted in the cleanup?

File: qe_metrics/cli.py [hunk 4]
```diff
@@ -35,15 +36,20 @@
...
+    ) as jira, orm.db_session:
+        # TODO: Run a cleanup of the database to remove old entries
+
+        for product in database.Products.from_file(products_file=products_file):
             for severity, query in product.queries.items():
-                # TODO: Execute Jira query and populate the database
-                pass  # TODO: Remove this line once the code is implemented
-    # TODO: Run a cleanup of the database to remove old entries
+                _logger.info(f'Executing Jira query for "{product.name}" with severity "{severity}"')
+                database.JiraIssues.create_update_issues(
```
Comment (myakove) [inline qe_metrics/cli.py:47]: I guess `create_update_issues` can raise exception, we need to catch it and continue to the next product. (and have a proper log that some product failed)

File: qe_metrics/libs/jira.py [hunk 5]
```diff
@@ -17,9 +17,21 @@ def __init__(self, config_file: str) -> None:
             config_file (str): Path to the yaml file holding database and Jira configuration.
         """
         self.logger = get_logger(name=self.__class__.__module__)
-
         self.jira_config = parse_config(config_file)["jira"]
 
+    def __enter__(self) -> "Jira":
+        return self
```
Comment (myakove) [inline qe_metrics/libs/jira.py:23]: `__enter__` should call `self.connection`

In context manager, I expect to have "working" object while I'm in the context, meaning I have a connection once I'm inside the context.

Same for `Database` class

File: qe_metrics/utils/issue_utils.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+from datetime import date, datetime
+
+from jira import Issue
+
+from simple_logger.logger import get_logger
+from typing import TYPE_CHECKING, List
+
+if TYPE_CHECKING:
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:9]: What this mean?

File: qe_metrics/utils/issue_utils.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    from qe_metrics.libs.database import Database
+
+LOGGER = get_logger(name=__name__)
+
+
+def check_customer_escaped(issue: Issue) -> bool:
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:15]: `is_customer_escaped_issue` maybe ?

File: qe_metrics/utils/issue_utils.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    Check if the issue is customer escaped.
+
+    Args:
+        issue (Issue): Jira Issue
+
+    Returns:
+        bool: True if the issue is customer escaped, False otherwise.
+    """
+    try:
+        return float(issue.fields.customfield_12313440) > 0
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:26]: Can you set a mapping of name: field and work with it?
`customfield_12313440` means nothing.

File: qe_metrics/utils/issue_utils.py (hunk 7 above)
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:15]: should be in Jira class.

File: qe_metrics/utils/issue_utils.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+def format_issue_date(date_str: str) -> date:
+    """
+    Format the date in the format YYYY-MM-DD.
+
+    Args:
+        date_str (str): Date string from Jira.Issue
+
+    Returns:
+        date: Date object for database insertion
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:39]:
```suggestion
        datetime.date: Date object for database insertion
```

File: qe_metrics/utils/issue_utils.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    return datetime.strptime(date_str, "%Y-%m-%dT%H:%M:%S.%f%z").date()
+
+
+def update_existing_issue(existing_issue: "Database.JiraIssues", issue: Issue, severity: str) -> "Database.JiraIssues":
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:44]: should be in `Database` class

File: qe_metrics/utils/issue_utils.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            LOGGER.info(
+                f'Updating issue "{issue.key}" in database: "{field}" changed from "{current_value}" to "{new_value}"'
+            )
+            setattr(existing_issue, field, new_value)
+            existing_issue.date_record_modified = datetime.now()
+
+    return existing_issue
+
+
+def delete_closed_issues(
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:74]: should move to Jira or Database class, need to check where it's fit more.

Also as I wrote in my previous comment, I don't think we need to delete them, let the cleanup do it

File: tests/unittests/conftest.py [hunk 12]
```diff
@@ -1,5 +1,8 @@
+from unittest.mock import MagicMock
```
Comment (myakove) [inline tests/unittests/conftest.py:1]: Why do you use `unittest` ?

**Review by myakove — CHANGES_REQUESTED** [review]

File: qe_metrics/libs/database.py [hunk 13]
```diff
@@ -127,3 +134,44 @@ class JiraIssues(DB_CONNECTION.Entity):  # type: ignore
         customer_escaped = orm.Required(bool)
         date_created = orm.Required(date)
         last_updated = orm.Required(date)
+        date_record_modified = orm.Required(datetime, default=datetime.now)
```
Comment (rnetser) [inline qe_metrics/libs/database.py:137]: why do we need both this and `last_updated`?

File: qe_metrics/libs/database.py (hunk 2 above)
Comment (rnetser) [inline qe_metrics/libs/database.py:159]: please use arg=val
motivation:
1. allows easier refactor (not using positional args)
2. more readable


please add this check to pre-commit:
- https://github.com/RedHatQE/openshift-python-wrapper/blob/main/.pre-commit-config.yaml#L18
- https://github.com/RedHatQE/openshift-python-wrapper/blob/main/.flake8#L52

File: qe_metrics/libs/database.py (hunk 3 above)
Comment (rnetser) [reply qe_metrics/libs/database.py:174]: moreover, i guess products would like to see these bugs (e.g a blocker would probably be addressed rather quickly but you still want this is the stats)

File: qe_metrics/utils/issue_utils.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+    Args:
+        issue (Issue): Jira Issue
+
+    Returns:
+        bool: True if the issue is customer escaped, False otherwise.
+    """
+    try:
+        return float(issue.fields.customfield_12313440) > 0
+    except (ValueError, AttributeError):
```
Comment (rnetser) [inline qe_metrics/utils/issue_utils.py:27]: in which case do we get `ValueError`?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: qe_metrics/libs/database.py (hunk 13 above)
Comment (calebevans) [reply qe_metrics/libs/database.py:137]: @rnetser `last_updated` maps to the "Updated" date in a Jira issue. The `date_record_modified` is the when the record is placed in the database or updated by our script. There is the potential to retrieve an issue that was "Updated" two weeks ago but was added to the database today. I was planning on using `date_record_modified` in the data retention clean up process. Do you think I should just use `last_updated`?

File: qe_metrics/libs/database.py (hunk 3 above)
Comment (calebevans) [reply qe_metrics/libs/database.py:174]: @rnetser @myakove sure, that makes sense. I will have to replace this with a function that sets the `status` to closed. The way the JQL queries are written, they will not find closed issues. So we will need a function to find issues in our db that are no longer being returned by the JQL queries (essentially what this does).

@rujutashinde just an FYI, you'll need to account for closed issues being in the db at least for 90 days. I will follow up after I fix the code.

File: qe_metrics/utils/issue_utils.py (hunk 6 above)
Comment (calebevans) [reply qe_metrics/utils/issue_utils.py:9]: @myakove `from qe_metrics.libs.database import Database` will cause a circular import but `Database` is used in the type annotations. This is essentially saying, only import if pre-commit hooks are running to check type annotations.

File: qe_metrics/utils/issue_utils.py (hunk 14 above)
Comment (calebevans) [reply qe_metrics/utils/issue_utils.py:27]: @rnetser Good point. This was essentially if that custom field returns a string that can't be converted to float. I'll remove the check for `ValueError`

File: qe_metrics/libs/database.py (hunk 1 above)
Comment (rujutashinde) [reply qe_metrics/libs/database.py:155]: @myakove I feel some check (either in the PR process or code) has to be there to ensure the additional ` issuetype` (story/task/epic/etc) are not added in the DB. Even if the responsibility lies with the user to provide the query, there should be some validation on the tool side. The check in the code seems to be a simple check instead of adding this in the CI-pipeline .

File: qe_metrics/libs/database.py (hunk 3 above)
Comment (calebevans) [reply qe_metrics/libs/database.py]: @myakove @rnetser I have discussed this with @rujutashinde, and we have decided to mark the "status" of these issues as `stale` in this instance, rather than deleting them.

**Comment by calebevans** [comment]
Note -- Support for Python 3.12 was removed because the [ORM only supports up to 3.11](https://blog.ponyorm.org/2023/09/26/pony-orm-release-0-7-17/). Using 3.12 seemed to cause some issues with using `lambda`.

File: qe_metrics/libs/database.py [hunk 15]
```diff
@@ -22,14 +21,13 @@ def __init__(self, config_file: str, verbose: bool) -> None:
...
-
+        orm.set_sql_debug(debug=verbose)
         self.config_file = config_file
         self.db_config = parse_config(path=self.config_file)["database"]
-        self.bind_local_db_connection() if self.db_config.get("local") else self.bind_remote_db_connection()
-        self.DB_CONNECTION.generate_mapping(create_tables=True)
 
     def __enter__(self) -> "Database":
+        self.bind_local_db_connection() if self.db_config.get("local") else self.bind_remote_db_connection()
+        DB_CONNECTION.generate_mapping(create_tables=True)
```
Comment (myakove) [inline qe_metrics/libs/database.py:30]: Maybe add logger success before returning.

File: qe_metrics/libs/orm_database.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+from pyaml_env import parse_config
+
+from qe_metrics.utils.general import verify_queries
+from qe_metrics.utils.general import format_issue_date
+
+DB_CONNECTION: orm.Database = orm.Database()
+
+
+class JiraIssuesEntity(DB_CONNECTION.Entity):
```
Comment (myakove) [inline qe_metrics/libs/orm_database.py:16]: JiraIsueEntity and ProductEntity should contain only DB data.
All other functions should be under Jira/Database classes.

File: qe_metrics/libs/orm_database.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    issue_key = orm.Required(str)
+    title = orm.Required(str)
+    url = orm.Required(str)
+    project = orm.Required(str)
+    severity = orm.Required(str)
+    status = orm.Required(str)
+    customer_escaped = orm.Required(bool)
+    date_created = orm.Required(date)
+    last_updated = orm.Required(date)
+    date_record_modified = orm.Required(datetime, default=datetime.now)
```
Comment (rnetser) [inline qe_metrics/libs/orm_database.py:32]: why is this needed?

File: qe_metrics/libs/orm_database.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+    @staticmethod
+    def mark_stale_issues(
```
Comment (rnetser) [inline qe_metrics/libs/orm_database.py:35]: should be in utils

File: tests/unittests/libs/test_database.py [hunk 19]
```diff
@@ -1,53 +1,166 @@
...
+        "last_updated": datetime.date(2023, 12, 31),
+    }
+    actual_values = {
+        "title": jira_issues[0].title,
+        "severity": jira_issues[0].severity,
+        "status": jira_issues[0].status,
+        "customer_escaped": jira_issues[0].customer_escaped,
+        "last_updated": jira_issues[0].last_updated,
+    }
+    assert actual_values == expected_values, f"{actual_values} != {expected_values}"
```
Comment (rnetser) [inline tests/unittests/libs/test_database.py:54]:
```suggestion
    assert actual_values == expected_values, f"actual: {actual_values} != expected: {expected_values}"
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: qe_metrics/libs/orm_database.py (hunk 17 above)
Comment (calebevans) [reply qe_metrics/libs/orm_database.py:32]: https://github.com/RedHatQE/qe-metrics/pull/59#discussion_r1542911790

File: qe_metrics/utils/issue_utils.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        current_issues (List[Issue]): A list of current Jira issues for a product.
+        db_issues ("JiraIssuesEntity"): A list of JiraIssuesEntity objects already in the database.
+        product ("ProductsEntity"): The product object to which the issues belong.
+    """
+    current_issue_keys = {issue.key for issue in current_issues}
+    db_issue_keys = {db_issue.issue_key for db_issue in db_issues if db_issue.product.name == product.name}
+    closed_issue_keys = db_issue_keys - current_issue_keys
+
+    for db_issue in db_issues:
+        if db_issue.issue_key in closed_issue_keys and db_issue.status != "obsolete":
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:42]: can you save and reuse `obsolete` as `OBSOLETE_STR`?

File: qe_metrics/utils/issue_utils.py [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+                f'Updating issue "{new_issue_data.key}" in database: "{field}" changed from "{current_value}" to "{new_value}"'
+            )
+            setattr(existing_issue, field, new_value)
+
+
+def create_update_issues(
+    issues: List[Issue],
+    product: "ProductsEntity",
+    severity: str,
+    jira_server: str,  # type: ignore  # noqa: F821
```
Comment (myakove) [inline qe_metrics/utils/issue_utils.py:75]: `# type: ignore  # noqa: F821` why this is needed?

File: qe_metrics/utils/product_utils.py [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    Args:
+        products_file (str): Path to the yaml file holding product names and their queries
+
+    Returns:
+        List[Dict[Any, Any]]: A list of dictionaries that hold the product and its queries
+    """
+    products_dict = parse_config(path=products_file)
+    products = []
+    for name, queries in products_dict.items():
+        verify_queries(queries_dict=queries)
```
Comment (myakove) [inline qe_metrics/utils/product_utils.py:21]: `verify_queries` raise, I think we want to continue to the next products if one fail.

**Review by myakove — CHANGES_REQUESTED** [review]

File: qe_metrics/utils/issue_utils.py (hunk 21 above)
Comment (calebevans) [reply qe_metrics/utils/issue_utils.py:75]: Must have been leftover from something else. Removed.

File: qe_metrics/utils/issue_utils.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    current_issue_keys = {issue.key for issue in current_issues}
+    db_issue_keys = {db_issue.issue_key for db_issue in db_issues if db_issue.product.name == product.name}
+    closed_issue_keys = db_issue_keys - current_issue_keys
+
+    for db_issue in db_issues:
+        if db_issue.issue_key in closed_issue_keys and db_issue.status != OBSOLETE_STR:
+            db_issue.logger.info(f'Marking issue "{db_issue.issue_key}" for product {product.name} as {OBSOLETE_STR}')
+            db_issue.status = OBSOLETE_STR
+
```
Comment (rnetser) [inline qe_metrics/utils/issue_utils.py:47]: orm.commit is needed here, right?

the functinon changes entities in the db so it should also be responsible for committing them

File: qe_metrics/utils/product_utils.py [hunk 24]
```diff
@@ -18,7 +21,10 @@ def products_from_file(products_file: str) -> List[Dict[Any, Any]]:
     products_dict = parse_config(path=products_file)
     products = []
     for name, queries in products_dict.items():
-        verify_queries(queries_dict=queries)
-        products.append({"product": ProductsEntity.get(name=name) or ProductsEntity(name=name), "queries": queries})
+        try:
+            verify_queries(queries_dict=queries)
+            products.append({"product": ProductsEntity.get(name=name) or ProductsEntity(name=name), "queries": queries})
+        except ValueError as err:
+            LOGGER.error(f"Error occurred parsing queries for product {name}: {err}")
```
Comment (myakove) [inline qe_metrics/utils/product_utils.py:28]: TODO:
since this is running application and we do not monitor the logs it is a good idea to have notification to slack on errors.

File: qe_metrics/utils/issue_utils.py [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+        ("status", new_issue_data.fields.status.name),
+        ("customer_escaped", new_issue_data.is_customer_escaped),
+        ("last_updated", format_issue_date(new_issue_data.fields.updated)),
+    ]
+    for field, new_value in fields:
+        if (current_value := getattr(existing_issue, field)) != new_value:
+            LOGGER.info(
+                f'Updating issue "{new_issue_data.key}" in database: "{field}" changed from "{current_value}" to "{new_value}"'
+            )
+            setattr(existing_issue, field, new_value)
```
Comment (rnetser) [inline qe_metrics/utils/issue_utils.py:70]: same here

File: qe_metrics/utils/product_utils.py (hunk 24 above)
Comment (calebevans) [reply qe_metrics/utils/product_utils.py:28]: @myakove Do you want me to add a TODO comment, or does this work? https://github.com/RedHatQE/qe-metrics/issues/71

I figure Slack will be used for more than just this error message, so an issue seems better suited.

**Review by rnetser — CHANGES_REQUESTED** [review]

File: qe_metrics/utils/issue_utils.py (hunk 25 above)
Comment (rnetser) [reply qe_metrics/utils/issue_utils.py:71]: orm.commit() should be called here as well

**Review by rnetser — CHANGES_REQUESTED** [review]

File: qe_metrics/utils/product_utils.py (hunk 24 above)
Comment (rnetser) [reply qe_metrics/utils/product_utils.py:28]: issue is enough; resolving

**Review by rnetser — APPROVED** [review]
